lib/mongoid/indexable/specification.rb

```diff
@@ -63,39 +63,73 @@ def name
 
       private
 
-      # Normalize the spec, in case aliased fields are provided.
+      # Normalize the spec in place, in case aliased fields are provided.
       #
       # @api private
       #
       # @example Normalize the spec.
-      #   specification.normalize_key(name: 1)
+      #   specification.normalize_aliases!(name: 1)
       #
-      # @param [ Hash ] key The index key(s).
+      # @param [ Hash ] spec The index specification.
       #
       # @return [ Hash ] The normalized specification.
-      def normalize_key(key)
-        normalized = {}
-        key.each_pair do |name, direction|
-          normalized[klass.database_field_name(name).to_sym] = direction
+      def normalize_aliases!(spec)
+        return unless spec.is_a?(Hash)
+
+        spec.transform_keys! do |name|
+          klass.database_field_name(name).to_sym
         end
-        normalized
       end
 
-      # Normalize the index options, if any are provided.
+      # Normalize the index options in place.
       #
       # @api private
       #
       # @example Normalize the index options.
-      #   specification.normalize_options(unique: true)
+      #   specification.normalize_options!(unique: true)
       #
-      # @param [ Hash ] opts The index options.
+      # @param [ Hash ] options The index options.
       #
       # @return [ Hash ] The normalized options.
-      def normalize_options(opts)
-        options = {}
-        opts.each_pair do |option, value|
-          options[MAPPINGS[option] || option] = value
+      def normalize_options!(options)
+
+        options.transform_keys! do |option|
+          option = option.to_sym
+          MAPPINGS[option] || option
+        end
+
+        %i[weights wildcard_projection].each do |key|
+          normalize_aliases!(options[key])
```

Comment:
Can you explain why `recursuve_normalize_conditionals` is not called here?

Reply:
Because these options do not support any nesting; their field hash is required to be one level.

Reply:
I see that. In general I think I would prefer a more general solution, i.e. one that is not specific to current server syntax. I am okay with specifying `weights`, `wildcard_projection` and `partial_filter_expression` but I think in general it would be a good idea to automatically recurse N-levels deep in any hash, in case the server syntax changes. This also would make for a slimmer and more easily readable recursive implementation.

Reply:
The two things I think that should change in favor of a more general solution:
1. always recurse N-levels deep
2. don't specify `$and`, `$or`, `$nor`, take any arbitrary operator (string beginning with $) into account